File: rheology_shear/plotting/shear_plot.m
ref_dir = "extracted_data/";
data_dir = "data/";

figure()
hold on

% colors = ["#9a090c",
% "#2d16bd",
% "#7b00b8",
% "#e93172",
% "#4b94de",
% "#d96ff2",
% "#ec8633",
% "#55c743"];
ms = 12;
lw = 5;

n_c = 5;
n_bot = 2;
n_top = 2;
n_tot = n_c + n_bot + n_top;
colors = ice(n_tot);
colors = colors(n_bot:end-n_top, :);
colors = flipud(colors);
beta_bright = 0.3;
beta_ref = -0.4;

% ref_colors = oslo(6);
% ref_colors = ref_colors(3:end-1, :);
% ref_colors = matter(n_tot-2);
% ref_colors = ref_colors(n_bot:end-n_top, :);
% ref_colors = flipud(ref_colors);

ref_colors = colors;

val = 0.1;
black = [val, val, val];


% dat = readmatrix(data_dir + "shear_stress_1.csv");
n_runs = 3;
[phi, dat, var_dat] = average_data(data_dir, n_runs);

std_err = 2*var_dat / sqrt(n_runs);

handles = [];
for i=1:5
  % h = errorbar(phi, 1+dat(:,i), std_err(:,i), Color=colors(i+3), Marker='square', MarkerFaceColor=colors(i+3));
  h = plot(phi, 1+dat(:,i), Color=brighten(colors(i, :), beta_bright), Marker='square', MarkerFaceColor=brighten(colors(i, :), ...
    beta_bright), MarkerSize=ms, LineWidth=lw);
  handles = [handles, h];
end

i=1;
ref_handles = [];
for n_ref = [12,42,162]
  filename = ref_dir + num2str(n_ref) + "-bead.csv";
  data = readmatrix(filename);
  phi = data(:,1);
  eta = data(:,2);
  h = plot(phi, eta, Color=brighten(ref_colors(i, :), beta_ref), Marker="diamond", MarkerFaceColor=brighten(ref_colors(i, :), beta_ref), ...
    LineStyle="--", MarkerSize=ms, MarkerEdgeColor="none", LineWidth=lw-2);
  ref_handles = [ref_handles, h];
  i = i +1;
end

phi_grid = linspace(0, 0.5, 200);
eta_func = @(p) (1+1.5*p.*(1+p.*(1+p-2.3*p.^2))) ./ (1 - p.*(1+p.*(1+p-2.3*p.^2)));
eta_analytic = eta_func(phi_grid);

h_ladd = plot(phi_grid, eta_analytic, '-k');

% temp = readmatrix("temp.csv");
% plot(temp(:,1), 1+temp(:,2), '--k');

ylim([0.5, 7.5]);
xlim([0, 0.5])

box on
daspect([0.5, 7, 1])
grid on

xlabel("$\phi$")
ylabel("$\eta_r$")
% handles = [h_ladd, ref_handles, handles];
handles = [handles, ref_handles, h_ladd];
% legend(handles, ["Ladd (1990)", "m=12 (ref.)", "m=42 (ref.)", "m=162 (ref.)", "m=12", "m=42", "m=162", "m=642", "m=2562"], Location="northwest", NumColumns=2)
legend(handles, ["m=12", "m=42", "m=162", "m=642", "m=2562", "m=12 (ref.)", "m=42 (ref.)", "m=162 (ref.)",  "Ladd (1990)"], Location="northwest", NumColumns=2)

xticks(0:0.1:0.5)
yticks(1:7)

function [phi, avg_dat, var_dat]  = average_data(data_dir, n_avg)


dat = readmatrix(data_dir + "shear_stress_1.csv");
all_dat = zeros([3, size(dat)] - [0,0,1]);

for i=1:n_avg
fname = data_dir + "shear_stress_" + num2str(i) + ".csv";
dat = readmatrix(fname)
phi = dat(:,1);
dat = dat(:,2:end);
all_dat(i, :, :) = dat;
end

avg_dat = mean(all_dat, 1);
var_dat = var(all_dat, 0, 1);

avg_dat = squeeze(avg_dat);
var_dat = squeeze(var_dat);
end

File: rheology_shear/plotting/shear_plot.py
from pathlib import Path
import numpy as np
import matplotlib.pyplot as plt
import logging

# Module-level logger
logger = logging.getLogger(__name__)

REF_DIR = Path("extracted_data")
DATA_DIR = Path("data")

N_RUNS = 3
N_C = 5
N_BOT = 2
N_TOP = 2
MARKER_SIZE = 12
LINE_WIDTH = 5
BETA_BRIGHT = 0.3  # brighten plotted series
BETA_REF = -0.4  # darken reference series
X_MIN, X_MAX = 0.0, 0.5
Y_MIN, Y_MAX = 0.5, 7.5
XTICKS = np.arange(0.0, 0.51, 0.1)
YTICKS = np.arange(1, 8, 1)
OUT_FIG = Path("comparison.png")


def brighten(rgb: np.ndarray, beta: float) -> np.ndarray:
    """
    MATLAB-like brighten:
      beta > 0 -> brighten toward white
      beta < 0 -> darken toward black
    """
    rgb = np.asarray(rgb, dtype=float)
    if beta >= 0:
        out = rgb + beta * (1.0 - rgb)
    else:
        out = rgb * (1.0 + beta)
    return np.clip(out, 0.0, 1.0)


def get_ice_colors(n_tot: int, n_bot: int, n_top: int) -> np.ndarray:
    """
    Attempt to use the 'ice' colormap from tsipkens/cmap.
    If unavailable, fall back to a perceptual matplotlib cmap.
    Slice [n_bot : n_tot - n_top] and flipud, as in MATLAB code.
    """
    # Try to import the optional 'cmap' package
    colors = None
    try:
        import cmap  # type: ignore

        ice = getattr(cmap, "ice", None)
        if callable(ice):
            colors = ice(n_tot)
    except Exception:
        colors = None

    if colors is None:
        # Fallback: use matplotlib's "viridis" (good perceptual properties)
        cm = plt.get_cmap("viridis")
        idx = np.linspace(0.0, 1.0, n_tot)
        colors = cm(idx)[:, :3]  # drop alpha if present

    colors = colors[n_bot : n_tot - n_top, :]
    colors = np.flipud(colors)
    return colors


def average_data(data_dir: Path, n_avg: int):
    """
    Python port of the MATLAB average_data function.
    Assumes CSV files: shear_stress_{i}.csv with columns [phi, series...].
    Returns:
      phi: (N,)
      avg_dat: (N, M)
      var_dat: (N, M)
    """
    # Read first to discover shape
    first = np.loadtxt(data_dir / "shear_stress_1.csv", delimiter=",")
    phi = first[:, 0]
    dat = first[:, 1:]  # N x M
    N, M = dat.shape
    all_dat = np.zeros((n_avg, N, M), dtype=dat.dtype)
    all_dat[0] = dat

    # Load remaining runs
    for i in range(2, n_avg + 1):
        fname = data_dir / f"shear_stress_{i}.csv"
        arr = np.loadtxt(fname, delimiter=",")
        if arr.shape[0] != N:
            raise ValueError(f"{fname} has {arr.shape[0]} rows, expected {N}.")
        all_dat[i - 1] = arr[:, 1:]

    avg_dat = np.mean(all_dat, axis=0)
    var_dat = np.var(all_dat, axis=0, ddof=0)
    return phi, avg_dat, var_dat


def main():
    # figure & axes
    fig, ax = plt.subplots(figsize=(8, 5), constrained_layout=True)

    n_tot = N_C + N_BOT + N_TOP
    colors = get_ice_colors(n_tot, N_BOT, N_TOP)
    ref_colors = colors.copy()

    # Load averaged data
    phi, dat, var_dat = average_data(DATA_DIR, N_RUNS)
    std_err = (
        2.0 * var_dat / np.sqrt(N_RUNS)
    )  # computed but not plotted (to match MATLAB)

    # Plot the 5 main series: 1 + dat(:, i)
    handles = []
    for i in range(5):
        c = brighten(colors[i, :], BETA_BRIGHT)
        (h,) = ax.plot(
            phi,
            1.0 + dat[:, i],
            marker="s",
            markerfacecolor=c,
            markeredgecolor="none",
            markersize=MARKER_SIZE,
            linewidth=LINE_WIDTH,
            color=c,
        )
        handles.append(h)

    # Reference curves: m = 12, 42, 162
    ref_handles = []
    for i, n_ref in enumerate([12, 42, 162]):
        fname = REF_DIR / f"{n_ref}-bead.csv"
        ref = np.loadtxt(fname, delimiter=",")
        phi_ref = ref[:, 0]
        eta_ref = ref[:, 1]
        c = brighten(ref_colors[i, :], BETA_REF)
        (h,) = ax.plot(
            phi_ref,
            eta_ref,
            linestyle="--",
            marker="D",
            markerfacecolor=c,
            markeredgecolor="none",
            markersize=MARKER_SIZE,
            linewidth=LINE_WIDTH - 2,
            color=c,
        )
        ref_handles.append(h)

    # Analytic curve (Ladd-like)
    phi_grid = np.linspace(0.0, 0.5, 200)
    eta_func = lambda p: (1 + 1.5 * p * (1 + p * (1 + p - 2.3 * p**2))) / (
        1 - p * (1 + p * (1 + p - 2.3 * p**2))
    )
    eta_analytic = eta_func(phi_grid)
    (h_ladd,) = ax.plot(phi_grid, eta_analytic, "-k")
    handles.extend(ref_handles)
    handles.append(h_ladd)

    # Axes formatting
    ax.set_ylim(Y_MIN, Y_MAX)
    ax.set_xlim(X_MIN, X_MAX)
    ax.grid(True)
    for spine in ax.spines.values():
        spine.set_visible(True)

    ax.set_xlabel(r"$\phi$")
    ax.set_ylabel(r"$\eta_r$")

    labels = [
        "m=12",
        "m=42",
        "m=162",
        "m=642",
        "m=2562",
        "m=12 (ref.)",
        "m=42 (ref.)",
        "m=162 (ref.)",
        "Ladd (1990)",
    ]
    ax.legend(handles, labels, loc="upper left", ncols=2)

    ax.set_xticks(XTICKS)
    ax.set_yticks(YTICKS)

    fig.savefig(OUT_FIG, dpi=200)
    logger.info("Wrote %s", OUT_FIG.resolve())


if __name__ == "__main__":
    main()
